From dca3a7fad76db623e206290b2e40db3e77743a8c Mon Sep 17 00:00:00 2001
From: nanli2x <nanli2x@intel.com>
Date: Sun, 27 Dec 2020 17:40:52 -0500
Subject: [PATCH] [PATCH] xen-blkback: set ring->xenblkd to NULL after
 kthread_stop()

When xen_blkif_disconnect() is called, the kernel thread behind the
block interface is stopped by calling kthread_stop(ring->xenblkd).
The ring->xenblkd thread pointer being non-NULL determines if the
thread has been already stopped.
Normally, the thread's function xen_blkif_schedule() sets the
ring->xenblkd to NULL, when the thread's main loop ends.

However, when the thread has not been started yet (i.e.
wake_up_process() has not been called on it), the xen_blkif_schedule()
function would not be called yet.

In such case the kthread_stop() call returns -EINTR and the
ring->xenblkd remains dangling.
When this happens, any consecutive call to xen_blkif_disconnect (for
example in frontend_changed() callback) leads to a kernel crash in
kthread_stop() (e.g. NULL pointer dereference in exit_creds()).

This is XSA-350.
Reported-by: default avatardefault avatarOlivier Benjamin <oliben@amazon.com>
Reported-by: default avatardefault avatarPawel Wieczorkiewicz <wipawel@amazon.de>
Reviewed-by: default avatardefault avatarJuergen Gross <jgross@suse.com>
Signed-off-by: default avatardefault avatarPawel Wieczorkiewicz <wipawel@amazon.de>
Fixes: a24fa22c ("xen/blkback: don't use xen_blkif_get() in xen-blkback kthread")
Signed-off-by: default avatardefault avatarAuthor Redacted <security@xen.org>
Reviewed-by: default avatardefault avatarJulien Grall <jgrall@amazon.com>
Reviewed-by: default avatardefault avatarJuergen Gross <jgross@suse.com>
---
 drivers/block/xen-blkback/xenbus.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/block/xen-blkback/xenbus.c b/drivers/block/xen-blkback/xenbus.c
index d19adf1db1f1..8fbcccd2c5b2 100644
--- a/drivers/block/xen-blkback/xenbus.c
+++ b/drivers/block/xen-blkback/xenbus.c
@@ -263,6 +263,7 @@ static int xen_blkif_disconnect(struct xen_blkif *blkif)
 
 		if (ring->xenblkd) {
 			kthread_stop(ring->xenblkd);
+			ring->xenblkd = NULL;
 			wake_up(&ring->shutdown_wq);
 		}
 
-- 
2.17.1

